### ISSUE TITLE
Auto answering help needed (+ sending messages from server to clients)

### ISSUE THREAD

**Issue by DjoleDzele** [open-easyrtc/open-easyrtc#97]
First thank you for awesome tool, i'm currently looking through code and finding "audio only" demo very useful for my scenario, but maybe you could point me in right direction about what i had in mind, basically im planing on using open-easyrtc in such way that it will be connection with 1 raspberry pi (which will be server), and 10-15 other computers, all in local network. basically i would need somehow to make RPI server to accept call immediately without need to answering manually when somebody calls it, but also if there is already some kind of flag or boolean within the existing code which i would be able to use to show other users on html page that server is currently busy in active call with someone else and unable to accept other calls, any advice would be great

Edit: also since all the callers get random id strings on each reload, and i want them only to be able to call RPI server and not make calls between each other, is there a way to create fixed ID string for RPI server for example "1000" or something which would always be the same so that i can hardcode it to existing call function, so each time they start call it only goes toward RPI and nothing else

**Comment by krneticz** [comment]
@DjoleDzele, you should use usernames instead of temporary ids to filter and auto-accept calls. Use easyrtc.setUsername(username) to set a permanent username for the clients. After that, use easyrtc.setAcceptChecker  and accept (or reject) the call automatically based on the caller's username. Hope this helps!

**Comment by DjoleDzele** [comment]
I'll look into that thank you, i was thinking to try using some kind of query string in URL to set server id to always be the same based on it, and for clients to stay random numbers as they are now, since they will only be calling server and not each other either way, but that might be helpful ill try it

**Comment by DjoleDzele** [comment]
I managed to work it out following your suggestions, basically since the browser on RPI is the only one that should be receiving calls and also auto answering them, i added ID query parameter only to RPI's browser url address, to differentiate it from clients which have no query params, then inside `demo_audio_only.js` file i added function `easyrtc.setUsername("id from query");`, and then i use existing listener ( `easyrtc.setRoomOccupantListener("my callback function");` ) which listens for new connections (i assume) to run my callback function, in which it loops over all current "room occupants" and compares if any of them has key "username" and checks if its the same as the hardcoded value i already decided upon and set in query string:

```
let alreadySetServerCalling = false;

function myCallbackFn(roomName, occupants, isPrimary) {
    for (var rtcId in occupants) {

        if (!alreadySetServerCalling && occupants[rtcId].hasOwnProperty("username") && occupants[rtcId].username.toString() == '100') {
            alreadySetServerCalling = true;
            callButton.onclick = function (rtcId) {
                return function () {
                    performCall(rtcId);
                };
            }(rtcId);
        }
    }
}
```

and then as you mentioned automatic answering can be set inside `easyrtc.setAcceptChecker`, i basically did it by just commenting out some extra stuff (will look into some cleaner solution later)

```
easyrtc.setAcceptChecker(function (easyrtcid, callback) {
    // document.getElementById('acceptCallBox').style.display = "block";
    // if (easyrtc.getConnectionCount() > 0) {
    //     document.getElementById('acceptCallLabel').textContent = "Drop current call and accept new from " + easyrtcid + " ?";
    // }
    // else {
    //     document.getElementById('acceptCallLabel').textContent = "Accept incoming call from " + easyrtcid + " ?";
    // }
    var acceptTheCall = function (wasAccepted) {
        // document.getElementById('acceptCallBox').style.display = "none";
        if (wasAccepted && easyrtc.getConnectionCount() > 0) {
            easyrtc.hangupAll();
        }
        callback(wasAccepted);
    };
    // document.getElementById("callAcceptButton").onclick = function () {
    //     acceptTheCall(true);
    // };
    // document.getElementById("callRejectButton").onclick = function () {
    //     acceptTheCall(false);
    // };
    acceptTheCall(true);
});
```

so basically all in all this seems to be as i imagined :) 

The only thing that was bugging me these days that i couldn't find solution to, how to notify other clients when one client is currently in call with RPI server, is there any EventListener i could subscribe to which can immediately let other clients know if some other client Started the call and when call is Hangup, or i would need to create some custom Events on my own, any idea would be helpful. I looked through docs here on github and also here ( https://courses.cs.washington.edu/courses/cse481v/21sp/projects/team1/node_modules/open-easyrtc/docs/docs_in_html/server-api/easyrtc_default_event_listeners.js.html ) but didnt find something similar to that?

**Comment by krneticz** [comment]
There are two ways to go about this. You can catch easyrtcCmd events on the server and specifically look for msgType "answer" and "hangup" to catch established calls and then emit info about the call to all clients in the room. Another option is to catch call start and end events from the client side and send peer messages to all other peers in the room.

On the server:

```
var onEasyrtcCmd = function (connectionObj, msg, socketCallback, next) {
  var caller= connectionObj.getEasyrtcid();

  easyrtc.events.emitDefault(
    "easyrtcCmd",
    connectionObj,
    msg,
    socketCallback,
    next
  );

    if (msg.msgType == "hangup" || msg.msgType == "answer"){
      // notify peers here
    }
  
};

easyrtc.events.on('easyrtcCmd', onEasyrtcCmd);
```

Use connectionObj.generateRoomClientList(roomStatusopt, nullable, roomMapnullable, callback) to find peers in the room and send messages using easyrtc.events.emit.

On the client side, it's probably a bit simpler. Catch when the call is started or ended and use getRoomOccupantsAsArray(roomName) → {Array}
 and sendPeerMessage(destination, msgType, msgData, successCB, failureCB) to notify all the peers.

**Comment by DjoleDzele** [comment]
Great, i think this will be it! Thanks for big help!

**Comment by DjoleDzele** [comment]
@krneticz  Well some update after a while, had no clue this will be so complicated, maybe im missing some crazy details, the client solution you gave is pretty straightforward to setup and works very good, but i realized that server side solution is better suited for all possible scenarios, following your guide i am able to intercept `msgType` and based on that also get required `easyrtcid` and all connected users and room data, but for some unknown reasons it just seems impossible to receive any kind of message sent from server to any client, let alone to specific client, wherever i look on forums or docs those questions are vagely answered, im already running in circles for a while


basically from your answer above, i tried to emit from server any kind of message to any client to at least console log it, inside your `onEasyrtcCmd` function in `if (msg.msgType === 'setRoomApiField') {}` i added `easyrtc.events.emit("easyrtcMsg", connectionObj, 'test123', socketCallback, next)`, which uses all the same main variables for parameters as the ones in your "easyrtc.events.emitDefault". And on client side i added 

```
easyrtc.setServerListener( function(msgType, msgData, targeting) {
    console.log("The Server sent the following message " + msgData);
});
```

the only output i get is inside terminal console (not browser) saying `warning - EasyRTC: [easyrtc.audioOnly][H5kiCVKU8aVlTQeR] EasyRTC message received with invalid message format [MSG_REJECT_BAD_STRUCTURE]. 'test123'`

if i use `msg` variable instead of string `test123`, i get different error but still nothing received on client console
```
warning - EasyRTC: [easyrtc.audioOnly][EobwQFQv3to68Axf] EasyRTC message received without targetEasyrtcid or targetRoom {
  msgType: 'setRoomApiField',
  msgData: {
    setRoomApiField: {
      roomName: 'default',
      field: {
        mediaIds: {
          fieldName: 'mediaIds',
          fieldValue: { default: '{c3ba0548-3e2e-4069-a05e-b3d76dfba906}' }
        }
      }
    }
  }
}
```


tried adding `var socket = io.connect('http://' + document.domain + ':' + location.port);
easyrtc.useThisSocketConnection(socket);` to client js from [this link](https://stackoverflow.com/questions/33322644/easyrtc-data-exchange-through-websockets) since it seems similar issue, no change 


looked [here](https://groups.google.com/g/easyrtc/c/w9EY5e4LyIc) and [here](https://groups.google.com/g/easyrtc/c/b-mUUgJw5Ak/m/ShCf5lzGFHEJ), no help


from [this example](https://groups.google.com/g/easyrtc/c/NqpzQs2OrlY/m/ZAZNeio7BAAJ) i tried emmiting like `easyrtc.events.emit("easyrtcMsg", {data: 'some data'}, next);` and then receiving message on client with both
```
easyrtc.setServerListener( function(msgType, msgData, targeting) {
    alert("The Server sent the following message " + msgData);
});

easyrtc.addEventListener('easyrtcMsg', (eventName, data) => {
    console.log('eventName ===> ' + eventName);
    console.log(data);
});
```
 i only get error on client html page saying `addEventListener called with a bad event name = easyrtcMsg`


from [this](https://groups.google.com/g/easyrtc/c/F2KYEyeqgl4/m/iafLqhGsY04J) example, i used `easyrtc.events.emit("emitEasyrtcCmd", connectionObj, msg, {"senderEasyrtcid":"BLABLABLA"}, null, function(err){console.log("Hangup message sent!");});` and then it seems message menages somehow to get to client, i get nothing inside browser console but i get error embedded in client html saying `received unknown message type from server, msgType is [object Object]`, so i guess some advance


I also tried almost all Listeners and Emitters from regular docs but no advance... Honestly this is some of the most confusing situations and issues i had to deal with when it comes to coding for some reason, hopefully im missing some stupid detail or something obvious.

**Comment by krneticz** [comment]
@DjoleDzele Can you try sending with this code:

```
sendMessageToEasyrtcId: function(appObj, easyrtcId, msgType, msgData, callback){
    appObj.connection(easyrtcId, function(con_err, con){
      var mymsg = {
        msgType: msgType,
        msgData: msgData
      };

      easyrtc.events.emit("emitEasyrtcMsg", con, msgType, mymsg, null, function(err){
        if (err) {
          callback(err);
        } else {
          console.log('emitEasyrtcMsg : success :', easyrtcId);
          callback();
        }
      });
    });
  }
```

Use this function to get the current easyrtc id from username:

```
easyrtcidToUsername: function(appObj, easyrtcid, callback){
    appObj.connection(easyrtcid, function(con_err, con){
      if (!con_err) {
        var username = con.getUsername();
        callback(username);

      } else {
        callback({"error" : con_err});
      }
    });
  }
```

Hope this helps!

**Comment by DjoleDzele** [comment]
Thanks for your reply, concerning your `sendMessageToEasyrtcId` function, when i try to invoke it, i dont seem to have access to any kind of `appObj` to be able to set it as parameter, my intellisense in Webstorm is not seeing it anywhere either, only showing appObj when tried calling function with `connectionObj.appObj` like this for example 
```
if (msg.msgType === 'setRoomApiField') {
        let newUserID = connectionObj.getEasyrtcid();
        sendMessageToEasyrtcId(connectionObj.appObj, newUserID, msg.msgType, {data:12345}, ()=>{console.log('message sent')})
```
but my server then crashes saying 
```
appObj.connection(easyrtcId, function(con_err, con){
       ^

TypeError: Cannot read properties of undefined (reading 'connection')

```
also tried creating new instance of appObj class as its somewhat shown [here at the top](https://courses.cs.washington.edu/courses/cse481v/21sp/projects/team1/node_modules/open-easyrtc/docs/docs_in_html/server-api/pub.appObj.html), but that doesnt seem to be allowed.


is  my issue maybe that im trying to run all these commands inside the file `server_ssl.js`, are most of the things not accessible in it or?

**Comment by DjoleDzele** [comment]
Yup, it seems to have something to do with https vs http, i manage to get it to work in http now using your functions, assigning appObj to local _appObj and then using it when calling your function, but in https when trying all that i get cross-origin errors in browser. I got local value of `appObj` using
```
let _appObj;

// Start EasyRTC server
var rtc = easyrtc.listen(httpApp, socketServer, { 'stillAliveInterval': 0 }, function(err, rtcRef) {
    console.log("Initiated");

    rtcRef.events.on("roomCreate", function(appObj, creatorConnectionObj, roomName, roomOptions, callback) {
        console.log("roomCreate fired! Trying to create: " + roomName);

        _appObj = appObj;

        appObj.events.defaultListeners.roomCreate(appObj, creatorConnectionObj, roomName, roomOptions, callback);
    });
});
```

previously since i used demo server example for https it used to look like this so i had no clue about getting value of any appObj
```
var rtc = easyrtc.listen(httpApp, socketServer);
```

I'll look into it now

**Comment by DjoleDzele** [comment]
Well, thanks to that black magic code of yours, and after looking into `server.js` and `server_ssl.js` files and combining some parts from them, it seems like success now! To be honest no idea how you managed to figure out all that stuff or which docs where you looking into, but great help from your side, here is the final version of my `server_ssl.js` file for serving over https and being able to receive messages on client from server! i hope this is it now, if i notice something in the future i will update here since it will probably be useful for someone having similar troubles!

`server_ssl.js` :
```
// Load required modules
var https = require("https");     // https server core module
var fs = require("fs");        // file system core module
var express = require("express");   // web framework external module
var io = require("socket.io"); // web socket external module

// This sample is using the easyrtc from parent folder.
// To use this server_example folder only without parent folder:
// 1. you need to replace this "require("../");" by "require("open-easyrtc");"
// 2. install easyrtc (npm i open-easyrtc --save) in server_example/package.json

var easyrtc = require("../"); // EasyRTC internal module

// Set process name
process.title = "node-easyrtc";

// Setup and configure Express http server. Expect a subfolder called "static" to be the web root.
var httpApp = express();
httpApp.use(express.static(__dirname + "/audioRazglas/"));

// Start Express https server on port 8443
var webServer = https.createServer({
    key: fs.readFileSync(__dirname + "/certs/localhost.key"),
    cert: fs.readFileSync(__dirname + "/certs/localhost.crt")
}, httpApp);

// Start Socket.io so it attaches itself to Express server
var socketServer = io.listen(webServer, { "log level": 1 });

// Cross-domain workaround presented below:
socketServer.origins(function(origin, callback) {
    if (origin && ![
        'https://localhost:8443/',
        '*'
    ].includes(origin)) {
        return callback('origin not allowed', false);
    }
    callback(null, true);
});

// Start EasyRTC server
let _appObj;

var rtc = easyrtc.listen(httpApp, socketServer, { 'stillAliveInterval': 0 }, function(err, rtcRef) {
    console.log("Initiated");

    rtcRef.events.on("roomCreate", function(appObj, creatorConnectionObj, roomName, roomOptions, callback) {
        console.log("roomCreate fired! Trying to create: " + roomName);

        _appObj = appObj;

        appObj.events.defaultListeners.roomCreate(appObj, creatorConnectionObj, roomName, roomOptions, callback);
    });
});

// Listen on port 8443
webServer.listen(8443, function () {
    console.log('listening on https://localhost:8443');
});

//SENDING MESSAGE FROM SERVER TO CLIENT--------------------------

let callingAllowed = true;
let currentCallerID;

let getClientsListCallback = function (_empty, roomData) {
    let _clients = roomData['default']['clientList']
    for (const clientsKey in _clients) {
        console.log(`${clientsKey} --- ${_clients[clientsKey]}`)
    }
}

function onEasyrtcCmd(connectionObj, msg, socketCallback, next) {
    console.log(connectionObj.getEasyrtcid() + " " + msg.msgType)

    easyrtc.events.emitDefault(
        "easyrtcCmd",
        connectionObj,
        msg,
        socketCallback,
        next
    );

    if (msg.msgType === 'setRoomApiField') {
        let newUserID = connectionObj.getEasyrtcid();
        sendMessageToEasyrtcId(_appObj, newUserID, msg.msgType, {data:12345}, ()=>{console.log('message sent')})
    }

    if (callingAllowed) {
        if (msg.msgType === "offer") {
            callingAllowed = false;
            currentCallerID = connectionObj.getEasyrtcid();
            sendMessageToEasyrtcId(_appObj, currentCallerID, msg.msgType, {calling:true}, ()=>{console.log('message sent')})
            console.log(currentCallerID + ' IS CALLING')
        }
    }

    if (msg.msgType === "answer") {
        //...some test code for answering
        connectionObj.generateRoomClientList(null, null, getClientsListCallback);
    }
    if ((currentCallerID === connectionObj.getEasyrtcid()) && !callingAllowed && msg.msgType === "hangup") {
        console.log(currentCallerID + ' HANGUP');
        callingAllowed = true;
    }
}

function sendMessageToEasyrtcId(appObj, easyrtcId, msgType, msgData, callback){
    appObj.connection(easyrtcId, function(con_err, con){
        var mymsg = {
            msgType: msgType,
            msgData: msgData
        };

        easyrtc.events.emit("emitEasyrtcMsg", con, msgType, mymsg, null, function(err){
            if (err) {
                callback(err);
            } else {
                console.log('emitEasyrtcMsg : success :', easyrtcId);
                callback();
            }
        });
    });
}

function usernameToEasyrtcids(appObj, username, callback){
    appObj.getConnectionEasyrtcids(function(err, res){
        if (err) {
            callback({ error : err });

        } else {
            console.log('########### usernameToEasyrtcids : conn_rtc_ids :', res);

            if (res.length > 0) {
                var rtc_ids = [];
                for (var i = 0; i < res.length; i++) {
                    appObj.connection(res[i], function(con_err, con){
                        if (!con_err) {
                            var usr = con.getUsername();

                            if (usr === username){
                                rtc_ids.push(res[i]);
                            }

                            if (i === res.length - 1) {
                                if (rtc_ids.length === 0) {
                                    callback({error: "Cannot find easyrtcids"});
                                } else {
                                    callback(rtc_ids);
                                }
                            }
                        } else {
                            callback({error: con_err});
                        }
                    });
                }
            } else {
                callback({error: "Cannot find easyrtcids"});
            }
        }
    });
}

easyrtc.events.on('easyrtcCmd', onEasyrtcCmd);

//SENDING MESSAGE FROM SERVER TO CLIENT--------------------------
```

and inside client js file i only needed:
```
easyrtc.setServerListener( function(msgType, msgData, targeting) {
    console.log("The Server sent the following message " + JSON.stringify(msgData));
});
```

**Comment by krneticz** [comment]
@DjoleDzele glad it helped :)
